### PR TITLE
docs: cockroachdb storage backend configuration

### PR DESCRIPTION
**Pull request by dboslee**
Adds documentation for configuring cockroachdb as a storage backend.

This is the first enterprise only backend so I'm not certain what is the best way to indicate this is enterprise only.

### CODE REVIEW

File: docs/pages/reference/backends.mdx [hunk 1]
```diff
@@ -1467,3 +1467,92 @@ teleport:
...
+
+<Admonition
+  type="warning"
+  title="IMPORTANT"
+>
+  `CockroachDB` can currently only be used to store Teleport's internal database in a
+  highly-available way. This will allow you to have multiple auth servers in your
+  cluster for a high availability deployment, but it will not also store Teleport audit events
+  for you in the same way that [DynamoDB](#dynamodb) or
+  [Firestore](#firestore) will.
```
Comment (zmb3) [inline docs/pages/reference/backends.mdx:1486]: This confused me on first read. It sounded like we were saying "be careful - you can only use this for HA and not for non-HA"
```suggestion
  CockroachDB can currently only be used to store Teleport's cluster state.
  It cannot be used for Teleport's audit log in the same way that
  [DynamoDB](#dynamodb) or Firestore](#firestore) wan.
```

I removed the backticks from CockroachDB because it is just a product name. It is not a command, code snippet, etc.

File: docs/pages/reference/backends.mdx [hunk 2]
```diff
@@ -1467,3 +1467,92 @@ teleport:
...
+You must also enable change feeds in CockroachDB's cluster settings. Teleport
+will configure this setting itself if granted `SYSTEM MODIFYCLUSTERSETTING`.
+
+```
+SET CLUSTER SETTING kv.rangefeed.enabled = true;
+```
+
+For more details on deploying CockroachDB see [CockroachDB's deployment options](https://www.cockroachlabs.com/docs/stable/choose-a-deployment-option).
+
+For more details on configuring multi-region survival goals in CockroachDB see
```
Comment (zmb3) [inline docs/pages/reference/backends.mdx:1508]: "For more details" suggests that there are at least some details here, but I don't see anything about configuring multi-region survival goals.

I would rephrase this. Perhaps something along the lines of this:

> There are several ways to configure multi-region deployments of CockroachDB,
> the details of which are not in scope for this guide. To learn about how to
> configure multi-region survival goals, see ...

Comment (dboslee) [reply docs/pages/reference/backends.mdx]: Makes sense I updated this for both the deployment and survival goals here https://github.com/gravitational/teleport/pull/37880/commits/b42399a329254ca6d30e7cbfdef2599b8ca8cb58

**Comment by russjones** [comment]
@dboslee Is this PR ready for re-review?

**Comment by dboslee** [comment]
> @dboslee Is this PR ready for re-review?

yep

File: docs/pages/reference/backends.mdx [hunk 3]
```diff
@@ -1467,3 +1467,91 @@ teleport:
...
+protect access to CockroachDB in this configuration because that is where Teleport
+secrets like keys and user records will be stored.
+
+<Admonition
+  type="warning"
+  title="IMPORTANT"
+>
+  CockroachDB can currently only be used to store Teleport's cluster state.
+  It cannot be used for Teleport's audit log in the same way that
+  [DynamoDB](#dynamodb) or Firestore](#firestore) wan.
```
Comment (ptgott) [inline docs/pages/reference/backends.mdx:1484]:
```suggestion
  [DynamoDB](#dynamodb) or [Firestore](#firestore) can.
```

**Review by ptgott — APPROVED** [review]

File: docs/pages/reference/backends.mdx [hunk 4]
```diff
@@ -1467,3 +1467,91 @@ teleport:
...
+    # establishing a change feed conneciton.
+    #
+    # If your certificates are not stored at the default ~/.postgresql
+    # location, you will need to specify them with the sslcert, sslkey, and
+    # sslrootcert parameters.
+    change_feed_conn_string: postgresql://user_name@database-address/teleport_backend?sslmode=verify-full
+
+    # ttl_job_cron is an optional parameter which configures the interval at which CockroachDB will expire backend
+    # items based on their time to live. By default this is configured to run every
+    # 20 minutes. This is used by Teleport to cleanup old resources that are no longer
```
Comment (ptgott) [inline docs/pages/reference/backends.mdx:1553]:
```suggestion
    # 20 minutes. This is used by Teleport to clean up old resources that are no longer
```

**Review by ptgott — APPROVED** [review]

**Comment by dboslee** [comment]
@zmb3 could you giving this a second look when you get a chance?

File: docs/pages/reference/backends.mdx [hunk 5]
```diff
@@ -1429,3 +1429,90 @@ the location set the `data_dir` value within the Teleport configuration file.
 teleport:
   data_dir: /var/lib/teleport_data
 ```
+
+## CockroachDB
+
+Teleport can use [CockroachDB](https://www.cockroachlabs.com/) as a storage backend
+to achieve high available and survive regional failures. You must take steps to
```
Comment (zmb3) [inline docs/pages/reference/backends.mdx:1436]:
```suggestion
to achieve high availability and survive regional failures. You must take steps to 
```

File: docs/pages/reference/backends.mdx [hunk 6]
```diff
@@ -1429,3 +1429,90 @@ the location set the `data_dir` value within the Teleport configuration file.
...
+Teleport can use [CockroachDB](https://www.cockroachlabs.com/) as a storage backend
+to achieve high available and survive regional failures. You must take steps to 
+protect access to CockroachDB in this configuration because that is where Teleport
+secrets like keys and user records will be stored.
+
+<Admonition
+  type="warning"
+  title="IMPORTANT"
+>
+  CockroachDB can currently only be used to store Teleport's cluster state.
```
Comment (zmb3) [inline docs/pages/reference/backends.mdx:1444]: Starting to feel like we should have a matrix on this page that lists each backend and each use case (cluster state, audit log, session recordings, etc) to indicate which backends support which types of data.

File: docs/pages/reference/backends.mdx [hunk 7]
```diff
@@ -1429,3 +1429,90 @@ the location set the `data_dir` value within the Teleport configuration file.
...
+CockroachDB, see [CockroachDB's deployment options](https://www.cockroachlabs.com/docs/stable/choose-a-deployment-option).
+To learn about how to configure multi-region survival goals, see
+[multi-region survival goals](https://www.cockroachlabs.com/docs/stable/multiregion-survival-goals).
+
+To configure Teleport to use CockroachDB as a storage backend:
+- Configure all Teleport Auth Service instances to use the CockroachDB backend in the
+  `storage` section of `teleport.yaml` as shown below.
+- Deploy several Auth Service instances connected to the CockroachDB storage backend.
+- Deploy several Proxy Service instances.
+- Make sure that the Proxy Service instances and all Teleport agent services that
```
Comment (zmb3) [inline docs/pages/reference/backends.mdx:1477]: This isn't specific to CockroachDB in any way, is it?

File: docs/pages/reference/backends.mdx [hunk 8]
```diff
@@ -1429,3 +1429,90 @@ the location set the `data_dir` value within the Teleport configuration file.
...
+  setting populated with the address of a load balancer for Auth Service
+  instances.
+
+```yaml
+teleport:
+  storage:
+    type: cockroachdb
+
+    # conn_string is a required parameter. It is a PostgreSQL connection string used
+    # to connect to `CockroachDB` using the PostgreSQL wire protocol. Client
```
Comment (zmb3) [inline docs/pages/reference/backends.mdx:1488]:
```suggestion
    # to connect to CockroachDB using the PostgreSQL wire protocol. Client
```

File: docs/pages/reference/backends.mdx [hunk 9]
```diff
@@ -1429,3 +1429,90 @@ the location set the `data_dir` value within the Teleport configuration file.
...
+    # parameters may be specified using the URL. For a detailed list of available
+    # parameters see https://www.cockroachlabs.com/docs/stable/connection-parameter
+    #
+    # If your certificates are not stored at the default ~/.postgresql
```
Comment (zmb3) [inline docs/pages/reference/backends.mdx:1492]: Should we mention that we require mTLS? (Do we?)

**Review by zmb3 — APPROVED** [review]
